### PR TITLE
[STORM-3711] Fix ARM CI with wrong usage of "arm64-graviton2" resources

### PR DESCRIPTION
**Pull request by liusheng**
In # apache/storm/pull/3344, we have enabled ARM CI with all Storm
modules enabled and switched to use the "arm64-graviton2" resources, but
I have made a mistak that the ARM CI actually still run on x86 than
ARM even the CI job shows the label "arm64-graviton2", details can be
found in the CI job logs[1]. Because the "arm64-graviton2" requires
explicit virt: [lxd|vm] tag, see[2].

Addtionally, we have skip the increasing swap space statement in
intergration test because it will raise error on ARM. And we only
enabled openjdk11 for ARM jobs because the jdk8 is unsupported, see[3].

[1] https://travis-ci.org/github/apache/storm/jobs/741780900
[2] https://docs.travis-ci.com/user/multi-cpu-architectures#testing-on-multiple-cpu-architectures
[3] https://travis-ci.com/github/liusheng/storm/jobs/430934514

### CODE REVIEW

**Comment by liusheng** [comment]
Hi @Ethanlm ,
I am so sorry I have made a mistake, the https://github.com/apache/storm/pull/3344 has introduced a wrong usage of Travis "arm64-graviton2" resources, it shows the "arm64-graviton2" label on the job, but in the CI logs, we can find it actually use x86 resources. So here we need to fix.

**Comment by liusheng** [comment]
Hmm. looks like this change still cannot correctly use "arm64-graviton2" ARM resources in Travis. according to the [CI results](https://travis-ci.org/github/apache/storm/builds/742395332) of this change, it still use **x86** than **ARM** in the ARM job. that is because currently the "arm64-graviton2" is only available in **travis-ci.com**, not available in **travis-ci.org**, and Storm projects use **travis-ci.org**, please see the [Travis ci docs](https://docs.travis-ci.com/user/multi-cpu-architectures#multi-cpu-availaibility).
Addtionally, it is recomended to use **travis-ci.com** than **travis-ci.org** and the latter will be closed down in the future. please see: https://devops.stackexchange.com/questions/1201/whats-the-difference-between-travis-ci-org-and-travis-ci-com
So, maybe we can promote to migrate to use "travis-ci.com" in Storm project ? or there is already  a plan ?

**Comment by Ethanlm** [comment]
I wonder how PR build passed last time.

**Comment by Ethanlm** [comment]
Thanks for looking into it. I was not careful enough to see it actually ran on amd64 instead of arm64. Also thanks for the links. I will see if it makes sense to switch to travis-ci.com. It seems to be still on beta though

**Comment by liusheng** [comment]
Hi @Ethanlm ,
Thanks a lot for paying attention for this, it looks like now the Storm projects has been migrated to travis-ci.com, lets test this again.

**Comment by liusheng** [comment]
Hi @Ethanlm ,
Thanks for your previous help about ARM CI, currently, Storm CI has been migrated to use `travis-ci.com`, in this patch, I have fixed in the .travis.yml and made small changes to make the ARM CI can run successfully in `travis-ci.com`. Could you please help to review again ?  Thanks a lot.

**Comment by liusheng** [comment]
Hi @Ethanlm ,
Sorry to disturb you again, could you please help to take a look this PR ? This change try to fix the ARM CI of Strom to running on real `arm64-graviton2` resource. Thank you.

File: .travis.yml [hunk 1]
```diff
@@ -41,6 +40,48 @@ matrix:
   include:
     - arch: s390x
       jdk: openjdk11
+    - dist: focal
```
Comment (Ethanlm) [inline .travis.yml:43]: The default dist configured above we use is `trusty`. Should we upgrade the default dist to `focal`? (it can be a separate work)

File: .travis.yml [hunk 2]
```diff
@@ -41,6 +40,48 @@ matrix:
...
+      group: edge
+      arch: arm64-graviton2
```
Comment (Ethanlm) [inline .travis.yml:45]: Is it possible to use arm64 directly? 
https://docs.travis-ci.com/user/reference/overview/#linux-travisyml-examples

File: .travis.yml [hunk 3]
```diff
@@ -41,6 +40,48 @@ matrix:
   include:
     - arch: s390x
       jdk: openjdk11
+    - dist: focal
+      group: edge
```
Comment (Ethanlm) [inline .travis.yml:44]: What does `group` mean here?

**Review by Ethanlm — COMMENTED** [review]
Sorry again. It took me long to come back to this PR. Thanks for the good work.

I have a few questions in this PR.

It is nice that it moved to travis-ci.com. I didn't do anything about it. So I guess Apache made the migration.

File: .travis.yml (hunk 3 above)
Comment (liusheng) [reply .travis.yml:44]: sorry, this is unnecessary, my mistake.

File: .travis.yml (hunk 2 above)
Comment (liusheng) [reply .travis.yml:45]: Hmm,  we have tried using the common `arm64` resources, but many modules(especially Integration-Test) are unstable and easy to fail. the new added `arm64-graviton` is anounced a better performance ARM resources, so switch to use it. 
Maybe we can try back to use the `arm64` resources as we have migrated from `travis-ci.org` to `travis-ci.com` to see if the problems still occur.

File: .travis.yml (hunk 1 above)
Comment (liusheng) [reply .travis.yml:43]: Currently, the "arm64-graviton2" resources are only available in `Xenial`, `Bonic`, `Focal` dists of Ubuntu, please see the [Travis doc](https://docs.travis-ci.com/user/reference/overview#which-one-do-i-use)， I just picked a lastest version.

**Comment by liusheng** [comment]
Thank for your review @Ethanlm , let me try to use the "arm64" resources on ARM CI to see if it is OK now.

File: .travis.yml (hunk 1 above)
Comment (Ethanlm) [reply .travis.yml:43]: Yeah, I was thinking maybe we can simply this by updating https://github.com/apache/storm/blob/fe2f7102e244336e288d26f2dde8089198ee4c33/.travis.yml#L28

File: .travis.yml (hunk 2 above)
Comment (Ethanlm) [reply .travis.yml:45]: Can you link to the announcement? It looks like they all should work as indicated in the https://docs.travis-ci.com/user/reference/overview/#linux-travisyml-examples and I don't see performance concerns. 
The arm64 configuration seems more clean and simple

File: .travis.yml (hunk 1 above)
Comment (liusheng) [reply .travis.yml:43]: OK, let's me try

File: .travis.yml (hunk 2 above)
Comment (liusheng) [reply .travis.yml:45]: There is not annoucement, just be cause I have tested many times, the ARM CI jobs are very easy to fail due to timeout errors previously. I have tested again in the PR https://github.com/apache/storm/pull/3361, it looks like now the common `arm64` resource is OK to run the ARM CI job :), so let's switch to use the `arm64` resources

**Comment by liusheng** [comment]
Hi @Ethanlm ,
Thanks for your help, I have tried some times in https://github.com/apache/storm/pull/3361, now we can run X86 and ARM64 sucessfully by using `xenial` dist of Ubuntu, and now the common `arm64` resources is OK for ARM CI running, I have updated the changes to this PR, please help to review again. thank you.

File: integration-test/run-it.sh [hunk 4]
```diff
@@ -36,10 +36,13 @@ then
   chmod o+rx /home/travis
 fi
 list_storm_processes || true
-# increasing swap space so we can run lots of workers
-sudo dd if=/dev/zero of=/swapfile.img bs=4096 count=1M
-sudo mkswap /swapfile.img
-sudo swapon /swapfile.img
+if [ "$(uname -m)" != aarch64 ]; then
```
Comment (Ethanlm) [inline integration-test/run-it.sh:39]: Is this change still needed after the change to arm64 and xenial

File: integration-test/run-it.sh [hunk 5]
```diff
@@ -101,3 +104,7 @@ for i in {1..20} ; do
 done
 list_storm_processes
 mvn test -DfailIfNoTests=false -DskipTests=false -Dstorm.version=${STORM_VERSION} -Dui.url=http://localhost:8744
+
+# Kill all storm processes after tests running, otherwise the Travis CI(ARM) will still wait and then fail by
+# timeout error even though all the tests passed
+sudo pkill -9 -u storm
```
Comment (Ethanlm) [inline integration-test/run-it.sh:110]: Is this change still needed after the change to arm64 and xenial

**Review by Ethanlm — COMMENTED** [review]
Thanks for the great effort!

I'd like to validate a few changes that was introduced with arm64-graviton2 and see if they are still required. Please see the comments. Thanks

File: integration-test/run-it.sh (hunk 4 above)
Comment (liusheng) [reply integration-test/run-it.sh:39]: yes, there still have problems of swap support on Travis ARM64 resources, you can take a look the error: 
https://travis-ci.com/github/apache/storm/jobs/460842259

Comment (Ethanlm) [reply integration-test/run-it.sh:39]: Thanks for validating. 

For future reference, the log is 
```
+sudo dd if=/dev/zero of=/swapfile.img bs=4096 count=1M
1048576+0 records in
1048576+0 records out
4294967296 bytes (4.3 GB, 4.0 GiB) copied, 27.2215 s, 158 MB/s
+sudo mkswap /swapfile.img
Setting up swapspace version 1, size = 4 GiB (4294963200 bytes)
no label, UUID=ba655d1c-230e-4feb-a06a-e342c114d7eb
+sudo swapon /swapfile.img
swapon: /swapfile.img: insecure permissions 0644, 0600 suggested.
swapon: /swapfile.img: skipping - it appears to have holes.
The command "if [[ $(uname -m) != 's390x' ]]; then /bin/bash ./dev-tools/travis/travis-script.sh `pwd` $MODULES; fi" exited with 255.
```

File: integration-test/run-it.sh (hunk 5 above)
Comment (Ethanlm) [reply integration-test/run-it.sh:110]: Saw this in your test job  https://travis-ci.com/github/apache/storm/jobs/460903786
```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received
The build has been terminated
```
Is this what you saw before?

Comment (liusheng) [reply integration-test/run-it.sh:110]: yes, I have tried again in #3361, you can take a look the CI logs: https://travis-ci.com/github/apache/storm/jobs/460903786
at the end of the log, even all the intgration tests ran sucessfully, but the job didn't exit immediately and then failed due to timeout. 
```
0.11s$ rm -rf $HOME/.m2/repository/org/apache/storm
cache.2
store build cache
Done. Your build exited with 0.
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received
The build has been terminated

```

Comment (Ethanlm) [reply integration-test/run-it.sh:110]: Thanks

**Review by Ethanlm — APPROVED** [review]
+1

Thank you very much for getting this fixed.
